### DISCUSSION TITLE
Default value for combobox in JUCE alertWindow

### DISCUSSION THREAD

**Discussion by damiensellier** [RomanKubiak/ctrlr#328] [Q&A] [answered]
Hi,
Do you know if there's a way to auto select a default option (the one displayed) from a combobox array in a JUCE alertWindow?

I checked the [JUCE alertWindow documentation](https://docs.juce.com/master/classAlertWindow.html#a79bf3cd42182334c1ebf7923f1e3581e)  but I cannot find an answer about that.

This is the content of my AlertWindow:

```
--	Create Array for the Preset Bank DropDown
	saPresetType = StringArray()
	saPresetType:set(0, String("Unit A [1U Preset]"))
	saPresetType:set(1, String("Unit B [1U Preset]"))
	saPresetType:set(2, String("Unit C [1U Preset]"))
	saPresetType:set(3, String("Unit D [1U Preset]"))
	saPresetType:set(4, String("Unit A & Unit B [2U Preset]"))
	saPresetType:set(5, String("Unit C & Unit D [2U Preset]"))
	saPresetType:set(6, String("All Units [4U Preset]"))
	saPresetType:set(7, String("All Units & Routing [Config]"))



	modalWindow = AlertWindow("\Store Preset to DP4", "Please enter the Preset Type, the destination\n number and the name of your Preset.", AlertWindow.NoIcon)
	modalWindow:addButton("Cancel", 0, KeyPress(KeyPress.escapeKey), KeyPress())
	modalWindow:addButton("    OK    ", 1, KeyPress(KeyPress.returnKey), KeyPress())
	modalWindow:addComboBox("cmbPresetType", saPresetType, "Preset Type")
	modalWindow:addTextEditor("PrgrmValueEdit", tostring(PrgrmDest), "Preset destination number [0 to 49]", false)
	modalWindow:addTextEditor("PrgrmNameEdit", PrgrmName, "Rename As [Up to 16 characters long]", false)
	modalWindow:setModalHandler(WindowCallback)

	panel:getModulatorByName("SyxToDP4Pix"):getComponent():setProperty("componentVisibility", 1, false)
  	modalWindow:addCustomComponent(CompSyxToDP4Pix)

```

The first one is always ticked but there's probably another way to do it. 
Since there's no default value from JUCE, I was thinking about changing the order from the array order itself so as to put the current config first. But having the right option ticked/selected without changing the order would be more logical.

Let me know if you've already done something like that.

Thanks for your help ;)

Damien

![Store Preset to DP4 AlertWindow](https://user-images.githubusercontent.com/55164385/123570953-39854180-d797-11eb-9df6-b629d9819d97.png)

**Comment by Ted-juh** [comment]
Can't you just set the modulatorValue with all "false" flags?

panel:getModulatorByName("NameOfNiftyComboBox"):getComponent():setModulatorValue(1, false, false, false)
To set it to Unit B [1U preset]

**Reply by damiensellier** [reply]
That's a good idea but I tried and it didn't work. 
John's solution `modalWindow:getComboBoxComponent("cmbPresetType"):setSelectedItemIndex(globalCmbPresetType, 2)`
seems to be the way to go. 
Thanks ;)

**Comment by dnaldoog** [comment, marked as answer]
@damiensellier  Hi there Damien,

You need: <code>setSelectedItemIndex</code>and <code>getSelectedItemIndex</code> - at least that is how I have always done it! I am using <code>panel:getGlobalVariable</code> get and set in this case to remember the value last selected.

Also included in the sample panel is one way (using timers) to stop the AlertWindow popping up when two instances of the panel are opened. There is a hidden unprotected uiButton with function that fires a timer on load in a hidden layer. That delay stops the AlertWindow firing.

Also see here https://ctrlr.org/forums/topic/why-is-a-modal-box-similar-to-the-terminator/

[Modal Window_1.0.2021-06-28_19-27.panel.zip](https://github.com/RomanKubiak/ctrlr/files/6725626/Modal.Window_1.0.2021-06-28_19-27.panel.zip)


<pre>
--
-- Called when a modulator value changes
-- @mod   http://ctrlr.org/api/class_ctrlr_modulator.html
-- @value    new numeric value of the modulator
--
myMethod = function(--[[ CtrlrModulator --]] mod --[[ number --]], value --[[ number --]], source)
    --	Create Array for the Preset Bank DropDown
    local globalCmbPresetType = panel:getGlobalVariable(15)

    saPresetType = StringArray()
    saPresetType:set(0, String("Unit A [1U Preset]"))
    saPresetType:set(1, String("Unit B [1U Preset]"))
    saPresetType:set(2, String("Unit C [1U Preset]"))
    saPresetType:set(3, String("Unit D [1U Preset]"))
    saPresetType:set(4, String("Unit A & Unit B [2U Preset]"))
    saPresetType:set(5, String("Unit C & Unit D [2U Preset]"))
    saPresetType:set(6, String("All Units [4U Preset]"))
    saPresetType:set(7, String("All Units & Routing [Config]"))

    modalWindow =
        AlertWindow(
        "Store Preset to DP4",
        "Please enter the Preset Type, the destination\n number and the name of your Preset.",
        AlertWindow.NoIcon
    )
    modalWindow:addButton("Cancel", 0, KeyPress(KeyPress.escapeKey), KeyPress())
    modalWindow:addButton("    OK    ", 1, KeyPress(KeyPress.returnKey), KeyPress())
    modalWindow:addComboBox("cmbPresetType", saPresetType, "Preset Type")
    modalWindow:getComboBoxComponent("cmbPresetType"):setSelectedItemIndex(globalCmbPresetType, 2)
    modalWindow:addTextEditor("PrgrmValueEdit", tostring(PrgrmDest), "Preset destination number [0 to 49]", false)
    modalWindow:addTextEditor("PrgrmNameEdit", PrgrmName, "Rename As [Up to 16 characters long]", false)

    ret = modalWindow:runModalLoop()

    if ret == 1 then
        PrgrmDest = tonumber(modalWindow:getComboBoxComponent("cmbPresetType"):getSelectedItemIndex())
        panel:setGlobalVariable(15, PrgrmDest)
        panel:getLabel("d"):setText("You chose \187 " .. PrgrmDest + 1)
    end
    modalWindow:setVisible(false)
end

</pre>

----

![modalwindow](https://user-images.githubusercontent.com/76831683/123627818-66942b80-d84d-11eb-9956-e6d98e9e7da8.png)

**Reply by damiensellier** [reply]
Hi John @dnaldoog I tried your demo panel and it's perfect!  I will have to adapt it a little bit to my needs knowing that PrgrmDest (actually the destination memory #) is a different variable from the program name and value from the combo.
Yes, you are right, I know for the modal box that shows up every time you launch the panel, it's so boring. I checked the ctrlr.org forum about that and added timers and source filters as well and it's good to  go.

Regarding the demo panel you posted, can you please explain the strategy with the global variable and what's the **15** argument for? Is it just a random index for this global variable? 
`local globalCmbPresetType = panel:getGlobalVariable(15)`

When I launch your demo panel, the combo index is  set to 7, but 7 is not written anywhere to start with, where does it come from?

Can I just use the following, to pre-set the combo option beforhand? : 
`modalWindow:getComboBoxComponent("cmbPresetType"):setSelectedItemIndex(CmbIndex,2 )`

With **CmbIndex** as variable depending on the current config (set somewhere else)? CmbIndex could be pre-set from 0 to 7...

Thanks a lot for your help as usual. This kind of exemple should be available in a wiki page about alertWindows. I really think we should work on something like that asap as a definitive resource for current and future ctrlr developers. We have to tlk seriously about that all together.

Damien

**Reply by damiensellier** [reply]
EDIT It's working perfectly. Thanks @DnaldeT  and @dnaldoog for the help. Thread marked as solved.

**Reply by dnaldoog** [reply]
> Thanks a lot for your help as usual. This kind of exemple should be available in a wiki page about alertWindows. I really think we should work on something like that asap as a definitive resource for current and future ctrlr developers. We have to tlk seriously about that all together.
> 

Yes you are right!

I knew the answer, but I had to track it down in one of my panels and then I thought - I have answered this before in the old forum - but we forget, so rather than trying to remember everything, we should just be able to look it up somewhere! Big time saver.

**Reply by damiensellier** [reply]
Roman Kubiak should add an admin for the Wiki on this github and we need a thread to organise everything together. I don't know how to ask for this kind of credentials, unfortunately Roman is discreet.

**Reply by damiensellier** [reply]
@dnaldoog John, I opened an issue report about the wiki. Roman Kubiak is ok to go ahead. If you want to be involved in the project as admin just confirm in the following thread : https://github.com/RomanKubiak/ctrlr/issues/329

**Comment by dobo365** [comment]
Very good one that I will steal with pride and include in my guide (with your credits of course).
I have always tried to avoid modal windows in general programming so didn't try them in Ctrlr besides the provided utils.xxx
This thread made me realize that I don't have the possibility to do a Save As... in my panels.
You  take a preset and you can only Save it on itself.
So, very nice example. thx!

**Reply by damiensellier** [reply]
For sure, help yourself :) I did the same thing at first (plain simple backup/restore sysex dump) and thought it was very limiting. We have the capability to decompose, reconstruct a sysex message easily with ctrlr so it's very relevant to offer the possibility to save/load presets at a different location, and with a different Preset Name. This part is just a fragment of the modal box script and also involved everything related to the message modification itself obviously.
